politeiawww/politeiawww.go

```diff
@@ -858,6 +866,95 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
 	}
 }
 
+// updateBestBlock updates the cached best block.
+func (p *politeiawww) updateBestBlock(bestBlock uint64) {
+	p.bbMtx.Lock()
+	defer p.bbMtx.Unlock()
+	p.bestBlock = bestBlock
+}
+
+// getBestBlock returns the cached best block if there is an active websocket
+// connection to dcrdata. Otherwise, it requests the best block from politeiad
+// using the the decred plugin best block command.
+func (p *politeiawww) getBestBlock() (uint64, error) {
+	p.bbMtx.RLock()
+	bb := p.bestBlock
+	p.bbMtx.RUnlock()
+
+	// the cached best block will equal 0 if there is no active websocket
+	// connection to dcrdata, or if no new block messages have been received
+	// since a connection was established.
+	if bb == 0 {
+		return p.getBestBlockDecredPlugin()
+	}
+
+	return bb, nil
+}
+
+// resetPiDcrdataWSSubs is responsible for resetting the wsDcrdata connection
+// and making necessary changes to the required changes to the politeiawww
+// state so that the service continues to function properly during and after
+// the reconnection.
+func (p *politeiawww) resetPiDcrdataWSSubs() error {
+	// The cached best block is set to zero so that in the time between
+	// reconnection and receiving the first new block message, instead of
+	// using the old cached value, politeiad is queried for the best block.
+	p.updateBestBlock(0)
+
+	return p.wsDcrdata.reconnect()
+}
+
+// setupPiDcrdataWSSubs subscribes and listens to websocket messages from
+// dcrdata that are needed for pi.
+func (p *politeiawww) setupPiDcrdataWSSubs() error {
+	err := p.wsDcrdata.subToNewBlock()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+	listenerLoop:
+		for {
+			msg, ok := <-p.wsDcrdata.Receive()
+			if !ok {
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
+					log.Errorf("Unable to reconnect to dcrdata, no longer " +
+						"listening.")
+					break listenerLoop
```

Comment:
If dcrdata is down for a short period, I think you would want to continue to retry the reconnection.  What is the plan for connectivity issues that last more than an instant?

Reply:
Yes it does. You can see this in the reconnect method in wsdcrdata.go.